onnxruntime/core/providers/webnn/builders/impl/gru_op_builder.cc

```diff
@@ -0,0 +1,191 @@
+  }
+
+  emscripten::val outputs = model_builder.GetBuilder().call<emscripten::val>("gru", input, weight, recurrent_weight,
+                                                                             steps, hidden_size, options);
+
+  if (has_Y) {
+    model_builder.AddOperand(output_defs[0]->Name(), outputs[1]);
+  }
+  if (has_Y_h) {
+    model_builder.AddOperand(output_defs[1]->Name(), outputs[0]);
+  }
+
+  return Status::OK();
+}
+
+bool GruOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                                     const WebnnDeviceType /*device_type*/, const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger)) {
+    LOGS(logger, ERROR) << "Cannot get input's shape";
+    return false;
+  }
+  int32_t steps = static_cast<int32_t>(input_shape[0]);
+
+  if (input_defs.size() > 4 && input_defs[4]->Exists()) {
+    if (!Contains(initializers, input_defs[4]->Name())) {
+      LOGS(logger, ERROR) << "GRU: sequence_lens must be constant";
+      return false;
+    }
+
+    const auto& sequence_lens_tensor = *initializers.at(input_defs[4]->Name());
+    std::vector<int32_t> sequence_lens;
+    if (!ReadIntArrayFrom1DTensor(sequence_lens_tensor, sequence_lens, logger)) {
+      LOGS(logger, ERROR) << "Cannot read sequence lens tensor";
+      return false;
+    }
+    if (!std::all_of(sequence_lens.begin(), sequence_lens.end(),
+                     [steps](int32_t lens) -> bool { return steps == lens; })) {
+      LOGS(logger, ERROR) << "GRU: every sequence length must be equal to input shape[0]";
+      return false;
+    }
+  }
+
+  NodeAttrHelper helper(node);
+  if (helper.HasAttr("activations")) {
+    const auto activations = helper.Get("activations", std::vector<std::string>{"Sigmoid", "Tanh"});
+
+    if (activations.size() >= 4) {
+      if (activations[0] != activations[2] || activations[1] != activations[3]) {
+        LOGS(logger, ERROR) << "GRU: forward and reverse directions must have the same activations";
+        return false;
+      }
+    }
+
+    const InlinedHashSet<std::string> supported_activations = {"Relu", "Tanh", "Sigmoid"};
+    if (!std::all_of(activations.begin(), activations.end(),
+                     [&supported_activations](const std::string& activation) -> bool {
+                       return supported_activations.contains(activation);
+                     })) {
+      LOGS(logger, ERROR) << "GRU: activations must be one of Relu, Tanh, Sigmoid";
+      return false;
+    }
+  }
+
+  if (helper.Get("clip", std::numeric_limits<float>::max()) != std::numeric_limits<float>::max()) {
+    LOGS(logger, ERROR) << "GRU: clip is not supported";
+    return false;
+  }
+
+  if (helper.Get("layout", 0) != 0) {
+    LOGS(logger, ERROR) << "GRU: batchwise (layout == 1) is not supported";
+    return false;
+  }
+
+  return true;
+}
+
```

Comment:
Missing `HasSupportedInputsImpl` for all input operands data type check.

Reply:
Oh, is this the beginning of using `opSupportLimits` to fall back to the CPU EP?

Reply:
Right, I haven't started using `opSupportLimits` yet util it is completed implemented in Chromium.

Currently we use hard code list to check which types are supported.

Reply:
Indeed, I also have to reviewed @philloooo's CR still: https://github.com/webmachinelearning/webnn/pull/755/files